src/util/address.rs

```diff
@@ -15,129 +15,176 @@
 //! Support for ordinary base58 Bitcoin addresses and private keys
 //!
 
-use secp256k1::{self, Secp256k1};
+use std::str::FromStr;
+use std::string::ToString;
+
+use bitcoin_bech32::{self, WitnessProgram};
+use secp256k1::Secp256k1;
 use secp256k1::key::{PublicKey, SecretKey};
 
 use blockdata::script;
 use blockdata::opcodes;
 use network::constants::Network;
 use util::hash::Hash160;
-use util::base58::{self, FromBase58, ToBase58};
+use util::base58;
+use util::Error;
 
 /// The method used to produce an address
-#[derive(Copy, Clone, PartialEq, Eq, Debug)]
-pub enum Type {
-    /// Standard pay-to-pkhash address
-    PubkeyHash,
-    /// New-fangled P2SH address
-    ScriptHash
-}
-
-/// An address-related error
-#[derive(Clone, PartialEq, Eq, Debug)]
-pub enum Error {
-    /// Private key did not represent a valid ECDSA secret key
-    Secp(secp256k1::Error)
+#[derive(Clone, PartialEq, Debug)]
+pub enum Payload {
+    /// pay-to-pkhash address
+    PubkeyHash(Hash160),
+    /// P2SH address
+    ScriptHash(Hash160),
+    /// Segwit address
+    WitnessProgram(WitnessProgram),
 }
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq)]
 /// A Bitcoin address
 pub struct Address {
     /// The type of the address
-    pub ty: Type,
+    pub payload: Payload,
     /// The network on which this address is usable
     pub network: Network,
-    /// The pubkeyhash that this address encodes
-    pub hash: Hash160
 }
 
 impl Address {
     /// Creates an address from a public key
     #[inline]
     pub fn from_key(network: Network, pk: &PublicKey, compressed: bool) -> Address {
         Address {
-            ty: Type::PubkeyHash,
             network: network,
-            hash: if compressed {
-                Hash160::from_data(&pk.serialize()[..])
-            } else {
-                Hash160::from_data(&pk.serialize_uncompressed()[..])
-            }
+            payload: Payload::PubkeyHash(
+                if compressed {
+                    Hash160::from_data(&pk.serialize()[..])
+                } else {
+                    Hash160::from_data(&pk.serialize_uncompressed()[..])
```

Comment:
Should the library support uncompressed pubkey at all? This is legacy just like pay-to-pubkey that is also not supported as we hope it goes extinct.

Reply:
If somebody wanted to write code for pay-to-pubkey I'd support it actually...I encountered somebody with an old wallet not too long ago (in 2017) who had some pay-to-pub outputs in an old Electrum wallet, and I had to walk him through importing to Core because Electrum had dropped support for them.

Not that I feel too strongly about this ... I agree uncompressed keys suck.